presto-common/src/main/java/com/facebook/presto/common/block/ByteArrayBlockEncoding.java

```diff
@@ -49,16 +51,27 @@ public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceO
     public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceInput)
     {
         int positionCount = sliceInput.readInt();
-
-        boolean[] valueIsNull = decodeNullBits(sliceInput, positionCount).orElse(null);
-
         byte[] values = new byte[positionCount];
-        for (int position = 0; position < positionCount; position++) {
-            if (valueIsNull == null || !valueIsNull[position]) {
-                values[position] = sliceInput.readByte();
-            }
+
+        // Fast track if no nulls present
+        if (!sliceInput.readBoolean()) {
+            sliceInput.readBytes(wrappedBuffer(values));
+            return new ByteArrayBlock(0, positionCount, null, values);
         }
 
+        boolean[] valueIsNull = new boolean[positionCount];
+        int nullPositions = decodeNullBits(sliceInput, valueIsNull);
+        int nonNullPositions = positionCount - nullPositions;
+
+        // Read compact values array and redistribute to non-null positions
+        sliceInput.readBytes(wrappedBuffer(values), nonNullPositions * SIZE_OF_BYTE);
```

Comment:
Use `public abstract void readBytes(byte[] destination, int destinationIndex, int length);`. Creating many Slice objects could hurt performance and reliability.